orchagent/aclorch.cpp

```diff
@@ -3195,6 +3373,93 @@ void AclOrch::init(vector<TableConnector>& connectors, PortsOrch *portOrch, Mirr
     }
 
 
+    if (platform == VS_PLATFORM_SUBSTRING)
+    {
+        // For testing on VS the following values will be used.
+        m_switchMetaDataCapabilities[TABLE_ACL_USER_META_DATA_RANGE_CAPABLE] = "true";
+        m_switchMetaDataCapabilities[TABLE_ACL_USER_META_DATA_MIN] = "1";
+        m_switchMetaDataCapabilities[TABLE_ACL_USER_META_DATA_MAX] = "7";
+        m_switchMetaDataCapabilities[TABLE_ACL_ENTRY_ATTR_META_CAPABLE] = "true";
+        m_switchMetaDataCapabilities[TABLE_ACL_ENTRY_ACTION_META_CAPABLE] = "true";
+        m_metaDataMgr.populateRange(1,7);
+    }
+    else
+    {
+        // check switch capability of Metadata attribute, action and range.
+        // SAI_SWITCH_ATTR_ACL_USER_META_DATA_RANGE support and range values.
+        // SAI_ACL_ENTRY_ATTR_ACTION_SET_ACL_META_DATA
+        // SAI_ACL_ENTRY_ATTR_FIELD_ACL_USER_META
+
+        sai_status_t status = SAI_STATUS_SUCCESS;
+        sai_attr_capability_t capability;
+        uint16_t metadataMin = 0;
+        uint16_t metadataMax = 0;
+        m_switchMetaDataCapabilities[TABLE_ACL_USER_META_DATA_MIN] = "0";
+        m_switchMetaDataCapabilities[TABLE_ACL_USER_META_DATA_MAX] = "0";
+        m_switchMetaDataCapabilities[TABLE_ACL_USER_META_DATA_RANGE_CAPABLE] = "false";
+        m_switchMetaDataCapabilities[TABLE_ACL_ENTRY_ATTR_META_CAPABLE] = "false";
+        m_switchMetaDataCapabilities[TABLE_ACL_ENTRY_ACTION_META_CAPABLE] = "false";
+
+        status = sai_query_attribute_capability(gSwitchId, SAI_OBJECT_TYPE_SWITCH, SAI_SWITCH_ATTR_ACL_USER_META_DATA_RANGE, &capability);
+        if (status != SAI_STATUS_SUCCESS)
+        {
+            SWSS_LOG_WARN("Could not query ACL_USER_META_DATA_RANGE %d", status);
+        }
+        else
+        {
+            if (capability.get_implemented)
+            {
+                sai_attribute_t attrs[1];
+                attrs[0].id = SAI_SWITCH_ATTR_ACL_USER_META_DATA_RANGE;
+                sai_status_t status = sai_switch_api->get_switch_attribute(gSwitchId, 1, attrs);
+                if (status != SAI_STATUS_SUCCESS)
+                {
+                    SWSS_LOG_WARN("Could not get range for ACL_USER_META_DATA_RANGE %d", status);
+                }
+                else
+                {
+                    SWSS_LOG_NOTICE("ACL_USER_META_DATA_RANGE, min: %u, max: %u", attrs[0].value.u32range.min, attrs[0].value.u32range.max);
+                    m_switchMetaDataCapabilities[TABLE_ACL_USER_META_DATA_RANGE_CAPABLE] = "true";
+                    m_switchMetaDataCapabilities[TABLE_ACL_USER_META_DATA_MIN] = std::to_string(attrs[0].value.u32range.min);
+                    m_switchMetaDataCapabilities[TABLE_ACL_USER_META_DATA_MAX] = std::to_string(attrs[0].value.u32range.max);
+                    metadataMin = uint16_t(attrs[0].value.u32range.min);
+                    metadataMax = uint16_t(attrs[0].value.u32range.max);
+                }
```

Comment:
Do we need typecast here?

Reply:
the metadata values supproted by platforms are  1-7 , and 1-4095 bit. I dont think any platform would support a 32 bit for this field or any tcam would have 4 billion enties.

Reply:
i have changed this code to restrict the range to 0-4095